### PR TITLE
Split chain into pre and post authz

### PR DESCRIPTION
**Pull request by mjudeikis**
#### What type of PR is this?

Minimal config change to split authz chain into 2 steps. This helps when constructing custom authentication flows and you need to extract `DefaultBuildHandlerChainFromAuthz` and `DefaultBuildHandlerChainBeforeAuthz` 

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 1]
```diff
@@ -992,10 +992,24 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
...
 func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) http.Handler {
+	handler := DefaultBuildHandlerChainFromAuthz(apiHandler, c)
+	handler = DefaultBuildHandlerChainBeforeAuthz(handler, c)
+	return handler
+}
+
+func DefaultBuildHandlerChainFromAuthz(apiHandler http.Handler, c *Config) http.Handler {
+	handler := apiHandler
+	handler = filterlatency.TrackCompleted(handler)
+	handler = genericapifilters.WithAuthorization(handler, c.Authorization.Authorizer, c.Serializer)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:1007]: now you have these twice. Compare below.

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 2]
```diff
@@ -992,10 +992,24 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
 func BuildHandlerChainWithStorageVersionPrecondition(apiHandler http.Handler, c *Config) http.Handler {
 	// WithStorageVersionPrecondition needs the WithRequestInfo to run first
 	handler := genericapifilters.WithStorageVersionPrecondition(apiHandler, c.StorageVersionManager, c.Serializer)
-	return DefaultBuildHandlerChain(handler, c)
+	return DefaultBuildHandlerChainFromAuthz(handler, c)
 }
 
 func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) http.Handler {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:998]: can you add go doc for all of these? "Leaving code better than it was".

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 1 above)
Comment (mjudeikis) [reply staging/src/k8s.io/apiserver/pkg/server/config.go]: done

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 3]
```diff
@@ -992,16 +992,29 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
...
+	handler := DefaultBuildHandlerChainFromAuthz(apiHandler, c)
+	handler = DefaultBuildHandlerChainBeforeAuthz(handler, c)
+	return handler
+}
 
+// DefaultBuildHandlerChainFromAuthz builds the handler chain from authorization onwards.
+func DefaultBuildHandlerChainFromAuthz(apiHandler http.Handler, c *Config) http.Handler {
+	handler := apiHandler
 	handler = filterlatency.TrackCompleted(handler)
-	handler = genericapifilters.WithAuthorization(handler, c.Authorization.Authorizer, c.Serializer)
```
Comment (aojea) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:1002]: why the order change? does it matter?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/config.go:1002]: right, old order was correct. We want to measure how long authz takes.

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 4]
```diff
@@ -992,16 +992,29 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
 func BuildHandlerChainWithStorageVersionPrecondition(apiHandler http.Handler, c *Config) http.Handler {
 	// WithStorageVersionPrecondition needs the WithRequestInfo to run first
 	handler := genericapifilters.WithStorageVersionPrecondition(apiHandler, c.StorageVersionManager, c.Serializer)
-	return DefaultBuildHandlerChain(handler, c)
+	return DefaultBuildHandlerChainFromAuthz(handler, c)
 }
 
+// DefaultBuildHandlerChain builds the default handler chain for the API server.
+// Is split in 2 parts - before and after authorization.
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:999]: It is split.

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 3 above)
Comment (mjudeikis) [reply staging/src/k8s.io/apiserver/pkg/server/config.go:1002]: This handler adding for delegations order not clicking in my head every time :D fixed (I hope so)

**Comment by liggitt** [comment]
I'm not sure I understand the motivation here. In a custom scenario, wouldn't customization of the authorizer sufficient? It seems like this split makes it easier to make mistakes miswiring the handler chain by calling the incorrect methods or in an incorrect order

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 3 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/server/config.go:1002]: > We want to measure how long authz takes

I thought that was already instrumented. If not, instrumenting seems useful and possible without splitting the handler chain, right?

**Comment by mjudeikis** [comment]
> I'm not sure I understand the motivation here. In a custom scenario, wouldn't customization of the authorizer sufficient? It seems like this split makes it easier to make mistakes miswiring the handler chain by calling the incorrect methods or in an incorrect order

This is needed in the case where you want to roll your own authz machinery, but you still want to use existing pre-auth flow for user-info. Maybe we can construct names more clearly to imply these should not be called:

```
InternalDefaultBuildHandlerChainBeforeAuthz
InternalDefaultBuildHandlerChainFromAuthz
```

 and imply in the comment that  these if used explicitly should be used with knowing how they work, and if not sure use `DefaultBuildHandlerChain` ?

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 3 above)
Comment (mjudeikis) [reply staging/src/k8s.io/apiserver/pkg/server/config.go:1002]: this was mistake in my ordering (handler load order,) fixed now, so noop)

**Comment by mjudeikis** [comment]
I renamed the methods a bit to make them clearer. Anything else to easy your concerns @liggitt ?

**Comment by liggitt** [comment]
> This is needed in the case where you want to roll your own authz machinery, but still want to use existing pre-auth flow for user-info

Wouldn't that look like replacing the authorization implementation with a custom implementation, not splitting handler chain building?

I still don't understand how these two new methods would be used properly. Splitting the handler chain arbitrarily at authz means anything after authz (like `filterlatency.TrackCompleted(handler)`) is just missing from your handler and you have to know to duplicate it / add it yourself. That holds true for any change made to that function.

**Comment by mjudeikis** [comment]
This stands true to what you say. It basically helps us to track fewer code carry patches as we can use the first part of (before authz and rest is our code). But if you think this carries too much of risk - let's close this as this is not such a bit code-diff/maintainenace. And maybe in 6-12m situations will be different and I have a better way to workaround this.

**Comment by deads2k** [comment]
A custom authorizer is already well supported by the library.  Is there something about your authorizer that means it requires different inputs or produces different outputs compared to the authorization interface built into the k8s.io/apiserver (generic not kube).

**Comment by mjudeikis** [comment]
> A custom authorizer is already well supported by the library. Is there something about your authorizer that means it requires different inputs or produces different outputs compared to the authorization interface built into the k8s.io/apiserver (generic not kube).

It does support it. And the output is totally the same. But we want to have just authz part different. All the part above is desired to be kept. So we get all metadata injected into the request context. And authz is just second-hander in the chain after than based on same metada as the core one. 
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/config.go#L1005-L1071

I see these changes somehow generating way more discussions than I expected. Should I just close it as this is not big diff to carry for us :/

**Comment by deads2k** [comment]
> It does support it. And the output is totally the same. But we want to have just authz part different.

To replace an authorizer, use https://github.com/kubernetes/kubernetes/blob/6c046796175de51e33290804a7665db5fa752a80/staging/src/k8s.io/apiserver/pkg/server/config.go#L114-L115

If the inputs and outputs are the same, then no changes are required in the handler chain.

> So we get all metadata injected into the request context

This suggests the inputs are different.  I'm intellectually interested in the inputs that are being analyzed that are not manageable via the requestinforesolver (injected here: https://github.com/kubernetes/kubernetes/blob/6c046796175de51e33290804a7665db5fa752a80/staging/src/k8s.io/apiserver/pkg/server/config.go#L189-L191)

But separating the default handling chain this way isn't a direction we're prepared to go.

**Comment by mjudeikis** [comment]
Ok :) of split is no-go will just close and maintain this in our fork.

**Pull request closed without merging**